doc/adrs/XXX_support-server-side-pagination-for-table.adoc

```diff
@@ -0,0 +1,108 @@
+= ADR-XXX Support Server-Side Pagination for Table Representation
+
+== Context
+
+The default pagination features of Material React Table are client-side.
+This means you have to have all of your data fetched and stored in the table all at once.
+This may not be ideal for large datasets.
+
+== Decision
+
+To address the limitations of client-side pagination, we will implement server-side pagination for Material React Table.
+This approach will allow us to fetch and display data more efficiently, especially for large datasets.
+The server-side pagination method will only be applied to table representations.
+For the use of tables as form widgets, client-side pagination will continue to be used.
+
+=== Material React Table
+
+To enable server-side pagination, we will set `manualPagination: true` in Material React Table.
+This will allow us to take control of pagination on the backend, ensuring that we can efficiently manage large datasets and accurately display the true total number of rows.
+The pagination must be managed within our own state, and listeners must be added to handle pagination state changes, allowing the backend to be called with the updated pagination data.
+
+[source, typescript]
```

Comment:
I really wanted to have this discussion with you. There are multiple ways to define the pagination each with specific tradeoffs. We should at least talk about our choice.

The two main strategies:

- page based (what you have described)
- cursor based

With a page based strategy I think it's easier to manipulate the API but it provides an inferior user experience. On top of that it offers inferior performances for some use cases. Let's consider that I'm iterating on everything starting from an object (which can even be the whole resource set).

´´´
aql:self.eAllContents()->subSequence(pageIndex * pageSize, pageSize)
´´´

In this case, we would still pretty much have to do an eAllContents() anyway, we wouldn't have a choice. I think that for this use case it would be faster with to start from an object with a cursor based pagination:

´´´
aql:self.eResource().getResourceSet().getEObject(cursor).nextElements(pageSize)
´´´

With something like that we would retrieve the current position in the iteration very quickly thanks to the cursor and then we would just have to find the ´pageSize´ next elements from there.

I don't see how we could let specifiers provide a performance efficient solution on this use case with a page based pagination.

On the other hand, because it's all about tradeoffs, if I want to return something like more complex, the dumb way to do it would be a less efficient with a cursor based pagination:

´´´
aql:self.friends->union(self.colleagues)->subSequence(pageIndex * pageSize, pageSize)
´´´

vs

´´´
aql:self.friends->union(self.colleagues)->subSequence(
  self.friends->union(self.colleagues)->indexOf(self.eResource().getResourceSet().getEObject(cursor)),
  pageSize
)
´´´

But we need to consider that in Java we would not have to compute the union twice, we would just keep it in a variable (which is not doable in raw aql). It would "just" require an additional ´indexOf´ which should not be that time consuming.

On the other hand, if one was able to write their expression as a computation with a starting point, it would be possible to do something like that:

´´´
aql: self.eResource().getResourceSet().getEObject(cursor).nextElementInTheUnionOfFriendsAndColleagues()
´´´

So if a specifier has the ability to compute not everything but instead all the next elements from a given point it should lower the amount of ´eAllContents´ performed because this will destroy the performances. I think one can write something like that because we could start an iterator at a given position with solething like this:

´´´
arrayList.listIterator(index)
´´´

I suspect that any serious Java service trying to support the pagination would not have to compute a collection but instead an iterator which would start at ´cursor´ or ´pageIndex * pageSize´ and return the next ´pageSize´ results. The question would now be what is more efficient: creating an iterator from a position or from the identifier of an object.

We definitely need to try to create an example of such an iterator and I have a good example in mind. Let's consider that I want to display a list of all the Papaya types with pagination. Thanks to a custom iterator I could very easily skip iterating on attributes, methods, parameters, and all their content. We need to try writing just the two versions of the tree iterator to compare the ease of use and performance.

Reply:
I'm not event sure that the code of the iterator would change, it's just the code of the method used to return the tree iterator which would have to change for page or cusor based pagination.

We would also need to think about other use cases from our community to see if we are not forgetting some use cases.

Reply:
I see another use case to take into account, the fact that we want to display the total number of results available.

Reply:
It's a very concern to take into account, we may have to iterate on all results anyway.

Here is an example of what I can come up with for the pagination with both strategies: https://gist.github.com/sbegaudeau/04dd722cd934a496e3d2dea9ea2a64a6

Reply:
But if we want to return the total count, we may be required to iterate on all potential results anyway which would make the difficulty in retrieving the starting point with a page based approach moot since we would have to use the custom iterator for the whole resource set anyway.

We could this way end up with pretty much requiring everyone to create a custom iterator (if they want good performances) to skip useless results but to always run the iterator on the whole resource set to have all results and thus compute the number of results.

It can be efficient (for example, in the case of my quick example, I can skip everything which is not a type and contained types) but it can't be perfect in the sense that we will need to iterate on more than just what we will return.

Reply:
So if we want to return the total count, we will have to iterate on everything to compute the length of the data returned. If we have to iterate on everything, we can keep having multiple rowDescriptions and execute all the `semanticCandidatesExpression` and we can simply compute both the `subList(start, end)` and the `.size()` by ourselves.

So in the end, and don't hesitate to correct me if you think I'm wrong, but I suspect that if we want to display the `totalCount`, we can just ask users for a list of elements in a semanticCandidatesExpression and we will do the rest because they won't be able to do better than us. It would be a simple API but it would condemn us to iterate on all considered elements at least once.

If I have 1 000 000 model elements and I want to display 100 000 of them in a table and the end user wants to see 10 of those at a time, during each refresh we would need at best to iterate on 100 000 elements and then execute all the other expression like a label expression etc on 10 elements.

I think we can live with this, what do you think?

We would have a simple API but not a perfect one in terms of performance.

Reply:
Okay, so to sum up, we can keep a simpler API page based .
We don't expose the variables linked to pagination, and we doing the split ourselves to send only the sub list that responds to pagination.
To do this, we use the complete list retrieved via rowDescription (which gives us the total number of rows).

It's ok for me.

Reply:
On the other hand if we want to maximuze performances, we would need the cursor based API with variables like cursor, direction (previous or next) and the size.

We would probably need additional aql services like EObject#preceding(int),  EObject#following(int) and EObject#next(direction, int) and some service to compute the a page object to be able write something like this

aql:self.eResource().getResourceSet().getEObject(cursor).next(direction, size)->asPage(...)

In both cases we need to find out the information needed in a page:

- size of the page?
- total count?
- hasNextPage
- hasPreviousPage

Some parameters could be optional like the total count for a maximum amount of performance.

Reply:
If we can get a good version of this ´toPage´ service, I think we could go for this approach